optimize host_group_vars and vars plugin loading

s-hertel opened the pull request:
##### SUMMARY
A targeted fix for host_group_vars efficiency based on changes that were in #79687 and #78814.

Using this [inventory script](https://gist.github.com/nitzmahone/2348775c22c6a6f0c35b1cb0e56ed097):

On devel:
```
$ time ansible-inventory -i ./inv.py --list --output /dev/null

real	1m14.201s
user	1m2.573s
sys	0m10.875s
```

On https://github.com/ansible/ansible/pull/79687:
```
$ time ansible-inventory -i ./inv.py --list --output /dev/null

real	0m7.197s
user	0m6.407s
sys	0m0.765s
```

The latest on this branch:
```
$ time ansible-inventory -i ./inv.py --list --output /dev/null

real	0m9.758s
user	0m8.792s
sys	0m0.918s
```

<details>
<summary>If you use a profiler and look at the specific function (`get_vars_from_path`), on devel the function accounts for ~78% of the runtime (it's the yellow one).</summary>
<div>
    <img src="https://user-images.githubusercontent.com/19572925/217396990-14d823de-44cc-439d-ad65-4d123f90e915.png" alt="image">
</div>
</details>

<details>
<summary>On #79687 it accounts for ~37% of the runtime (it's green).</summary>
<div>
    <img src="https://user-images.githubusercontent.com/19572925/217397089-6ad16dd4-236f-4be5-82c0-508be388a59b.png" alt="image">
</div>
</details>

<details>
<summary>The latest version on this branch it accounts for ~14% (it's blue).</summary>
<div>
    <img src="https://user-images.githubusercontent.com/19572925/234350489-e994da2f-84af-4f1c-a4c1-11c879da6a4a.png" alt="image">
</div>
</details>

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
vars plugins

s-hertel committed:
Improve host_group_vars efficiency:

* normalize the basedir with `os.path.realpath()` once and cache it
* cache missing paths/files
* reduce the calls to `isinstance`

Add a couple more general improvements in vars/plugins.py get_vars_from_path():

* call `PluginLoader.all()` once for vars plugins and reload specific
  plugins subsequently
* don't reload legacy/builtin vars plugins that are not enabled

Add a test for host_group_vars and legacy plugin loading

Co-authored-by: Matt Davis <mrd@redhat.com>

s-hertel committed:
changelog

s-hertel commented on changelogs/fragments/79945-host_group_vars-improvements.yml:
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+  - Cache host_group_vars after loading it once since it is stateless and limit the amount of work it needs to do every time it runs.
```
```suggestion
  - Cache host_group_vars after instantiating it once since it is stateless. Also limit the amount of repetitive work it needs to do every time it runs.
```

s-hertel committed:
Add a new is_stateless attribute to the vars plugin baseclass

update integration tests to be quieter and use the same test pattern

Fix deprecation and adjust test that didn't catch the issue (deprecation only occured when the value was False)

move realpath cache to host_group_vars (do not smuggle call state as instance data)

refactor under a single 'if cache:' statement

Call os.path.isdir instead of always calling os.path.exists first. Just call os.path.exists to differentiate between missing and non-directory.

remove call to super(VarsModule, self).get_vars()

use the entity name as the cache key instead of variable location

Remove isinstance checks and use a class attribute just in case any plugins are subclassing Host/Group

Replace startswith by checking index 0 of the name instead, since host/group names are required

nitzmahone commented on lib/ansible/plugins/vars/host_group_vars.py:
```diff
@@ -90,11 +99,16 @@ def get_vars(self, loader, path, entities, cache=True):
-                    opath = to_text(b_opath)
+                    opath = os.path.join(self._basedir, subdir)
+                    b_opath = to_bytes(opath)
+                    if cache and opath in NAK:
```
It's nice to avoid the extra layer of indent, but a single outer `if cache:` over all these  might be preferable in this case.

s-hertel replied:
Updated. I added another `continue` in the case of the files being found in the cache so I could keep the find + cache results at the same indentation level.

nitzmahone commented on lib/ansible/plugins/vars/host_group_vars.py:
```diff
@@ -90,11 +99,16 @@ def get_vars(self, loader, path, entities, cache=True):
+                        continue
                     else:
                         # no need to do much if path does not exist for basedir
                         if os.path.exists(b_opath):
```
Since we're trying to optimize the cached case, we could also ditch this (relatively) expensive call by collapsing the exists + isdir check into just `os.path.isdir` first, and then disambiguate the "missing or not a dir" in the `else` case.

s-hertel replied:
I replaced the nested `os.path.exists + os.path.isdir` with `if os.path.isdir/elif os.path.exists/else`.

nitzmahone commented on lib/ansible/plugins/vars/__init__.py:
```diff
@@ -38,4 +43,6 @@ def __init__(self):
-        self._basedir = basedir(path)
+        if path not in CANONICAL_PATHS:
+            CANONICAL_PATHS[path] = os.path.realpath(basedir(path))
+        self._basedir = CANONICAL_PATHS[path]
```
I always thought this overall pattern was weird, (smuggling call state as instance data), and at least for the current usage, for no apparent benefit. Caching aside, this `self._basedir` set technically invalidates the `stateless` thing. So maybe `host_group_vars` shouldn't call this via `super()`, and this cache should just be moved inline to `host_group_vars.get_vars` and both should use a local instead of `self._basedir`, or an instance method that provides access to the cache (since a `super()` call can't just change the `path` value to the `realpath`, which I think is really what we want for this case?).

bcoca replied:
'basedir' is mostly 'ported over' from when vars plugins were hardcoded plugins inside inventory and vars manager code. I always meant to clean it up, but was never a priority.

s-hertel replied:
I've just left `basedir` as-is for now (well, I added it to host_group_vars since it no longer calls super().get_vars()), but I moved the cache for the realpath into host_group_vars.

nitzmahone commented on lib/ansible/plugins/vars/host_group_vars.py:
```diff
@@ -78,10 +82,15 @@ def get_vars(self, loader, path, entities, cache=True):
         for entity in entities:
-            if isinstance(entity, Host):
+            entity_id = id(entity)
+            if entity_id in HOSTS or entity_id in GROUPS:
```
Still feels like there should be a cheaper heuristic for this than packratting id's. This is also potentially a little dangerous since this is basically using `id()` as a lighter-weight `WeakRef` surrogate. Python id's can get reused when an object is GC'd, so it's not impossible for an ID that was a `Host`  to show up later as a `Group` or vice-versa.

bcoca replied:
'basedir' is mostly 'ported over' from when vars plugins were hardcoded plugins inside inventory and vars manager code. I always meant to clean it up, but was never a priority.

bcoca replied:
internally we use host name and group name as unique identifiers everywhere else, we should keep that here as the host/group objects can be regenerated and have diff ids even for same host/group.

nitzmahone replied:
Since we've always been requiring specific types here anyway with the `isinstance`, and we fully control those types, we can actually do this a lot cheaper by either just doing `type(entity) is Host`, or if we want to allow for the possibility of subclasses coming back from a plugin, do something silly like add new class-level attrs on `Host` and `Group` like `base_type = Host`, then still the runtime cost is just `entity.base_type is Host` (makes the eval cost even cheaper at just an attr lookup and a reference compare). That way we don't have to cache anything, but the tight-loop cost is still very low.

s-hertel replied:
I added  the new class attr in an abundance of caution and removed the `isinstance` calls.

nitzmahone reviewed (changes requested):
Yeah, overall approach is good. A couple things to think about for safe + cheap ways to distinguish `Host` and `Group` without packratting id's or using `isinstance`, and maybe some stuff to move inline to the plugin (or consider deprecating `self._basedir` on vars plugins?). Once we decide and implement those things, this one's good for merge and backport from me...

bcoca commented:
We should also have a way to flush the cache,  at least for when we refresh inventory.

bcoca commented on lib/ansible/plugins/vars/host_group_vars.py:
```diff
@@ -55,33 +55,55 @@
 FOUND = {}  # type: dict[str, list[str]]
+NAK = set()
+HOSTS = set()
+GROUPS = set()
```
Should this cache be at the plugin level or at vars manager as a caller? (plugin + path + entity).

s-hertel replied:
Maybe that would be simpler. Let me try that for comparison's sake at least.

bcoca replied:
just so we have a few 3rd party to test against:
https://github.com/ansible-collections/community.sops/blob/main/plugins/vars/sops.py
https://github.com/manicminer/ansible-vars-plugin-aws/blob/master/vars_plugins/aws.py

tangent: this is just a simple example on writing vars plugins, several usable/testable ones. 
Aside from using for this ticket, I think we should incorporate a lot of this into our docs:
https://www.lilatomic.ca/posts/ansible_writing_facts_plugin/

s-hertel replied:
Even though I'm thinking now that variable manager would be a better place to do this in a generic way because of the other vars plugins, I'm not sure how to handle this for merge_hash behavior without worsening performance if there are a lot of host_vars/group_vars. To avoid accidentally mutating the cached variables, I think I'd need to deep copy the new variables before passing them to combine_vars. Keeping the cache at the plugin level allows loading the variables every time while caching some of the overhead to do that (or skipping it).

s-hertel replied:
@bcoca I added a commit to try to cover some of the documentation gaps.

s-hertel commented:
Current graph:

![image](https://user-images.githubusercontent.com/19572925/223561100-ae16c0c3-766b-4122-b30c-ebba6d2d8238.png)

bcoca commented on docs/docsite/rst/dev_guide/developing_plugins.rst:
```diff
@@ -477,62 +477,99 @@ Vars plugins
+    def run(self, name, vault_password=None):
+        pass # your code goes here
+
+Ansible 2.0 did not pass passwords to older plugins, so vaults were unavailable.
```
i think we can remove this line at this point

bcoca commented on docs/docsite/rst/dev_guide/developing_plugins.rst:
```diff
@@ -477,62 +477,99 @@ Vars plugins
+
+Ansible 2.0 did not pass passwords to older plugins, so vaults were unavailable.
+
+Since Ansible version 2.4, vars plugins only execute as needed when preparing to execute a task. This avoids the costly 'always execute' behavior that occurred during inventory construction in older versions of Ansible. Since Ansible version 2.10, vars plugin execution can be toggled by the user to run when preparing to execute a task or after importing an inventory source.
```
We can remove the first part of the phrase and just start at the current behavior since 2.10

bcoca commented on lib/ansible/plugins/vars/host_group_vars.py:
```diff
@@ -55,60 +55,92 @@
 FOUND = {}  # type: dict[str, list[str]]
+NAK = set()
+HOSTS = set()
+GROUPS = set()
```
the thing about adding the cache, is_stateless is now a lie, but used to reuse the cache ... should we rename the property?

s-hertel replied:
The caches are globals though, not tied to the instance.

s-hertel replied:
I renamed it to `cache_instance`, but feel free to bikeshed.

bcoca replied:
ah, missed that, but 'bikeshedding'TM is my game!

bcoca commented on lib/ansible/vars/plugins.py:
```diff
@@ -12,12 +12,45 @@
+cached_stateless_vars_plugins = {}
+
+
+def initialize_vars_plugin_caches():
```
I still think the cache should be in plugin loader, but we hit the fork problem, at least for collection based vars plugins enabled at task time.

bcoca commented on test/integration/targets/old_style_vars_plugins/deprecation_warning/vars.py:
```diff
@@ -2,7 +2,7 @@
 
 class VarsModule(BaseVarsPlugin):
-    REQUIRES_WHITELIST = False
+    REQUIRES_WHITELIST = True
```
~I thought we renamed this?~ we should update to use the newer `REQUIRES_ENABLED`

s-hertel replied:
This is testing the deprecation warning so I was going to delete it once the code can be deleted. We have a test for REQUIRES_ENABLED.

s-hertel committed:
rename is_stateless to cache_instance to make it more clear what it does

samccann commented on docs/docsite/rst/dev_guide/developing_plugins.rst:
```diff
@@ -477,62 +477,97 @@ Vars plugins
+           return variables
 
-This ``get_vars`` method just needs to return a dictionary structure with the variables.
+In addition to running many times, a vars plugin is instantiated by default every time it might run. Stateless plugins like ``host_group_vars`` can opt into reusing the same instance throughout a play by setting the ``cache_instance`` class attribute to True.
```
```suggestion
In addition to running many times, a vars plugin is instantiated by default every time it might run. Stateless plugins like ``host_group_vars`` can opt into reusing the same instance throughout a play by setting the ``cache_instance`` class attribute to ``True``.
```

samccann commented on docs/docsite/rst/dev_guide/developing_plugins.rst:
```diff
@@ -477,62 +477,97 @@ Vars plugins
+       cache_instance = True
 
-Legacy vars plugins are always loaded and run by default. You can prevent them from automatically running by setting ``REQUIRES_ENABLED`` to True.
+Vars plugins in non-collection search paths run by default without user-configuration. To prevent this, set the ``REQUIRES_ENABLED`` class attribute to True, and document that the legacy plugin needs to be enabled to run.
```
```suggestion
Vars plugins in non-collection search paths run by default without user-configuration. To prevent this, set the ``REQUIRES_ENABLED`` class attribute to ``True``, and document that the legacy plugin needs to be enabled to run.
```

samccann commented on docs/docsite/rst/dev_guide/developing_plugins.rst:
```diff
@@ -477,62 +477,97 @@ Vars plugins
+Historically, vars plugins have had a few changes in implementation and behavior.
+
+Vars plugins were partially implemented in Ansible 2.0 and rewritten to be fully implemented starting with Ansible 2.4. Vars plugins are supported by collections starting with Ansible 2.10.
+
```
Given we are on 2.15, do we need to keep the historical info here? Is it of value to anyone creating plugins today?

s-hertel replied:
I could delete one sentence:
```
Vars plugins were partially implemented in Ansible 2.0 and rewritten to be fully implemented starting with Ansible 2.4. 
```

But the historical example below still works for example and hasn't been deprecated. If ansible collections still support 2.9, I think it's worth keeping the differences since 2.9. (For the developer documentation specifically)

bcoca replied:
good point, should we take this as the opportunity to deprecate 1.0 vars plugins?

samccann replied:
yeah 2.9 is still relevant for sure. thanks!

s-hertel replied:
I was mistaken about the example below still working - we actually removed v1 vars plugins a while ago so I just removed that example. We never added developer documentation for (v2?) vars plugins using `get_host_vars`/`get_group_vars` instead of `get_vars`, so I added a deprecation for that.

samccann commented on docs/docsite/rst/plugins/vars.rst:
```diff
@@ -18,25 +18,29 @@ The :ref:`host_group_vars <host_group_vars_vars>` plugin shipped with Ansible en
+Vars plugins that run after parsing inventory, will execute for every host and group added to inventory for every inventory source directory. The groups 'all' and 'ungrouped' are always added to inventory.
 
-Starting in Ansible 2.10, vars plugins can be made to run at specific times. `ansible-inventory` does not use these settings, and always loads vars plugins.
+Before executing a task, vars plugins execute for every targetted entity for every inventory source directory and the playbook directory. The targetted entities are the group 'all', the hosts running the task, and the groups of the hosts running the task.
```
```suggestion
Before executing a task, vars plugins execute for every targeted entity for every inventory source directory and the playbook directory. The targeted entities are the group 'all', the hosts running the task, and the groups of the hosts running the task.
```

s-hertel commented on docs/docsite/rst/dev_guide/developing_plugins.rst:
```diff
@@ -477,62 +477,97 @@ Vars plugins
+
+Historically, vars plugins have had a few changes in implementation and behavior.
+
+Vars plugins were partially implemented in Ansible 2.0 and rewritten to be fully implemented starting with Ansible 2.4. Vars plugins are supported by collections starting with Ansible 2.10.
```
```suggestion
Vars plugins are supported by collections starting with Ansible 2.10.
```

s-hertel commented:
/azp run

s-hertel committed:
add plugin instance cache using the path to plugin loader

reduce loading stage option if a new instance isn't created

don't require a known subdir on PluginLoader instantiation for backwards
compatibility

rename attribute again

contain reading from/initializing cached instances to a plugin loader method

bcoca commented:
There could be an issue with precedence, since playbook (group|host)_vars > inventory (group|host)_vars, but since there is no precedence level in between i doubt it creates an actual issue .... lets just make a note to never add one in between .. or just add one in general.

s-hertel commented:
Do you mean for vars plugin options with `vars` in their config? I think it depends, because a vars plugin would have to call `set_options` itself. If it does that in `__init__`, we'd use that when getting the stage option, but if it's called in `get_vars` (using pre-existing entity vars for example) it wouldn't be set by the time we get the stage option.

Since it's opt-in to reuse the instance (and by extension, the stage option), I don't think this will change precedence for existing plugins. `host_group_vars` uses ini/env which should stay the same throughout.

Looks like I totally broke filters/tests in PluginLoader though, and I'm not sure what I did...

bcoca commented:
I was thinking that if we cache the path + entity combos we get 

- patha + groupa (cause inventory file is in patha)
- patha + groupa (cause playbook is in patha)

deduced into one call, this right now does not matter as the precedence of variables has nothing in between, but if tomorrow we do:

 - inventory vars plugin group (patha + group)
 - new directory for group (pathb + group)
 - playbook vars plugin group (patha + group)

Then the behavior between caching the dir or not, becomes a precedence issue.

But currently it helps with the useless x2 processing we used to do when inventory file was in same directory as the playbook.

s-hertel commented:
Oh, I see what you mean. This still does the 2x processing, but I was tempted to fix that. The last cache I added with the path is the path to the plugin, not the vars.

After I moved the cache into PluginLoader I realized it was much slower because `_resolve_step ` adds up even for legacy plugin names and we always loaded the config defs. So I changed it to cache the path and the original plugin load context so we could load the object without repeating those.

mkrizek commented:
!waffling needs_revision

s-hertel commented:
Here's a more minimal view using get_vars as the leaf node. It runs the same number of times.

<details>
<summary>devel</summary>
<div>
    <img src="https://user-images.githubusercontent.com/19572925/234357034-5d3ccecd-27b8-4f77-9243-1251a8be08d2.png" alt="image">
</div>
</details>

<details>
<summary>this branch</summary>
<div>
    <img src="https://user-images.githubusercontent.com/19572925/234357159-d1ff7284-0b3d-44f0-9693-f4111eab1e36.png" alt="image">
</div>
</details>

s-hertel committed:
Deprecate v2 vars plugins

s-hertel commented on docs/docsite/rst/dev_guide/developing_plugins.rst:
```diff
@@ -487,33 +487,88 @@ Vars plugins
+               variables.update(self.load_variables_from_source(loader, FOUND[cache_key]))
+           return variables
+
+In addition to running many times, a vars plugin is instantiated by default every time it might run. Stateless plugins like ``host_group_vars`` can opt into reusing the same instance throughout a play by setting the ``reuse_instance`` class attribute to ``True``.
```
Other attribute name suggestions: ``is_stateless``, ``use_cached_instance``, ``cache_instance``...

nitzmahone replied:
I'm +0.25 for `is_stateless` over `reuse_instance`, but either is fine with me...

s-hertel replied:
I renamed it back to `is_stateless`. It feels like more of a promise on the part of the vars plugin, than a description of the behavior.

nitzmahone commented on lib/ansible/inventory/host.py:
```diff
@@ -31,6 +31,7 @@
 
 class Host:
     ''' a single ansible host '''
+    base_type = 'Host'
```
Since this is always control-side code and we can use modern Python features, this would go quite a bit faster if it were an enum instead of a string- something we can get a singleton for and always safely sniff with `is`, as that's by far the cheapest operation we can do.

Right above this, define `class InventoryObjectType(Enum)` with members `HOST` and `GROUP`, then change the `Host`/`Group` class vars to use the enum value instead of the string, and the comparison logic below to use `is`.

s-hertel replied:
I updated it to use enums. I wanted to understand why it's more efficient, but using `==` with a string seems to be slightly faster than `is` with a singleton (but difference seems negligible).

nitzmahone commented on lib/ansible/plugins/vars/host_group_vars.py:
```diff
@@ -55,60 +55,92 @@
+        # realpath is expensive
+        if path not in CANONICAL_PATHS:
+            CANONICAL_PATHS[path] = os.path.realpath(basedir(path))
+        realpath_basedir = CANONICAL_PATHS[path]
```
This is a great place to use walrus, since we know this is a really hot code path- you can avoid consulting the path cache twice:

```suggestion
        # realpath is expensive
        if not (realpath_basedir := CANONICAL_PATHS.get(path)):
            CANONICAL_PATHS[path] = realpath_basedir = os.path.realpath(basedir(path))
```

s-hertel replied:
Adding the `get` here seems to be more expensive (using the test case from #79687), so I removed the extra path cache consultation by accessing the key immediately and handling the KeyError instead. Once the key is cached it makes sense that blindly accessing would be the most efficient.

<details>
<summary>some testing</summary>
<br>

```python
>>> def approach_1(container):
...     if 'something' not in container:
...         container['something'] = 'value'
...     value = container['something']
... 
>>> def approach_2(container):
...     if not (value := container.get('something')):
...         container['something'] = value = 'value'
... 
>>> def approach_3(container):
...     try:
...         value = container['something']
...     except KeyError:
...         container['something'] = value = 'value'
... 
>>>
>>> container = {i: "value" for i in range(5000)}
>>> timeit.timeit('approach_1(container)', globals=globals())
0.04540829104371369
>>>
>>> container = {i: "value" for i in range(5000)}
>>> timeit.timeit('approach_2(container)', globals=globals())
0.055584517074748874
>>>
>>> container = {i: "value" for i in range(5000)}
>>> timeit.timeit('approach_3(container)', globals=globals())
0.04524283786304295
>>>
>>>
>>> def exacerbate(func, number):
...     def wrapper(container):
...         for i in range(number):
...             func(container)
...     return wrapper
... 
>>>
>>> container = {i: "value" for i in range(10)}
>>> cProfile.run('exacerbate(approach_1, 5000)(container)')
         5005 function calls in 0.003 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     5000    0.001    0.000    0.001    0.000 <stdin>:1(approach_1)
        1    0.000    0.000    0.000    0.000 <stdin>:1(exacerbate)
        1    0.002    0.002    0.003    0.003 <stdin>:2(wrapper)
        1    0.000    0.000    0.003    0.003 <string>:1(<module>)
        1    0.000    0.000    0.003    0.003 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}


>>> container = {i: "value" for i in range(10)}
>>> cProfile.run('exacerbate(approach_2, 5000)(container)')
         10005 function calls in 0.004 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     5000    0.002    0.000    0.002    0.000 <stdin>:1(approach_2)
        1    0.000    0.000    0.000    0.000 <stdin>:1(exacerbate)
        1    0.002    0.002    0.004    0.004 <stdin>:2(wrapper)
        1    0.000    0.000    0.004    0.004 <string>:1(<module>)
        1    0.000    0.000    0.004    0.004 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
     5000    0.001    0.000    0.001    0.000 {method 'get' of 'dict' objects}


>>> container = {i: "value" for i in range(10)}
>>> cProfile.run('exacerbate(approach_3, 5000)(container)')
         5005 function calls in 0.001 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     5000    0.000    0.000    0.000    0.000 <stdin>:1(approach_3)
        1    0.000    0.000    0.000    0.000 <stdin>:1(exacerbate)
        1    0.000    0.000    0.001    0.001 <stdin>:2(wrapper)
        1    0.000    0.000    0.001    0.001 <string>:1(<module>)
        1    0.000    0.000    0.001    0.001 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}


>>>
>>>
>>> container = {i: "value" for i in range(10)}
>>> timeit.timeit('exacerbate(approach_1, 5000)(container)', globals=globals())
255.56839252007194
>>>
>>> container = {i: "value" for i in range(10)}
>>> timeit.timeit('exacerbate(approach_2, 5000)(container)', globals=globals())
263.8696411601268
>>>
>>> container = {i: "value" for i in range(10)}
>>> timeit.timeit('exacerbate(approach_3, 5000)(container)', globals=globals())
213.75293880398385
```
</details>

bcoca commented on docs/docsite/rst/plugins/vars.rst:
```diff
@@ -47,6 +51,10 @@ You can also control vars plugin execution on a per-plugin basis for vars plugin
 
+.. note::
+
+   ``ansible-inventory`` is hardcoded to always execute vars plugins and does not use these configuration options.
```
it still uses the enabled  and other config, should we specify that this note only applies to 'execution time' configs

bcoca commented on lib/ansible/inventory/group.py:
```diff
@@ -55,6 +55,7 @@ def to_safe_group_name(name, replacer="_", force=False, silent=False):
 
 class Group:
     ''' a group of ansible hosts '''
+    base_type = 'Group'
```
this feels like re implementing `x.__class__`

s-hertel replied:
I'm using a class attribute instead of `__class__` in the (unlikely) case Group/Host is subclassed.

bcoca commented on lib/ansible/plugins/__init__.py:
```diff
@@ -141,3 +141,6 @@ def _no_options(self, *args, **kwargs):
         return self._function
+
+
+PLUGIN_INSTANCE_CACHE = {}  # type: dict[str, dict[str, AnsiblePlugin]]
```
this cache won't be updated correctly from forks, as the data will be lost once the fork is completed

s-hertel replied:
:hole: :rabbit2:

nitzmahone commented on lib/ansible/plugins/vars/host_group_vars.py:
```diff
@@ -55,60 +55,92 @@
 FOUND = {}  # type: dict[str, list[str]]
+NAK = set()
+HOSTS = set()
+GROUPS = set()
```
Since we should be reusing the instance in all cases now anyway, any reason not to just make these normal instance vars on our plugin instead of module-level state?

s-hertel replied:
Wouldn't that make the plugin not stateless? This is why I wanted to bikeshed the attr name, because I think vars plugins are going to opt in to the feature but keep using instance attributes, which will now persist throughout a play.

nitzmahone commented on lib/ansible/plugins/vars/host_group_vars.py:
```diff
@@ -55,60 +55,92 @@
+GROUPS = set()
+
+
+def load_found_files(loader, data, found_files):
```
Same here- any reason this needs to be a module-level global function instead of a plugin instance method?

s-hertel replied:
Fixed. I just made this a standalone function because I wasn't using any instance attributes.

bcoca reviewed (commented):
the 'in plugin' cache seems to have the issue of not persisting across forks, but will benefit if the plugin is run both at 'inventory time' (pre fork) and task time (in fork) but only in one direction.

rest of code lgtm

nitzmahone commented on lib/ansible/plugins/__init__.py:
```diff
@@ -141,3 +141,6 @@ def _no_options(self, *args, **kwargs):
         return self._function
+
+
+PLUGIN_INSTANCE_CACHE = {}  # type: dict[str, dict[str, AnsiblePlugin]]
```
Since we don't want people digging around in this, and we're only using it on one type right now, probably best to hide this as an (optional and `_` prefixed) instance var of PluginLoader instead of exposing it on the top-level module for all to ~abuse~ see, and to get rid of the top layer of indirection (so it's just `dict[str, AnsiblePlugin] | None`).

s-hertel replied:
Fixed, there is now just the internal `self._plugin_instance_cache` VarsLoader attribute.

nitzmahone commented on lib/ansible/plugins/loader.py:
```diff
@@ -274,6 +281,7 @@ def __setstate__(self, data):
 
         PATH_CACHE[class_name] = data.get('PATH_CACHE')
         PLUGIN_PATH_CACHE[class_name] = data.get('PLUGIN_PATH_CACHE')
+        PLUGIN_INSTANCE_CACHE[class_name] = data.get('PLUGIN_INSTANCE_CACHE')
```
I don't think we can safely serialize/deserialize these instances in the general case... Honestly, I think we can/should probably get rid of serialization of `PluginLoader` altogether, but that's a bigger conversation.

In this case, it'd probably be simplest if we just always ignore it on serialization and re-create as an empty dictionary on deserialization (maybe even in `__init__` as well?) and only ever consult or populate it when we know we're doing vars plugins.

s-hertel replied:
Fixed, the internal cache is re instantiated as an emtpy dict for vars plugins only in `__init__` (which is called just below in this method).

nitzmahone commented on lib/ansible/plugins/loader.py:
```diff
@@ -295,6 +303,7 @@ def __getstate__(self):
             _searched_paths=self._searched_paths,
             PATH_CACHE=PATH_CACHE[self.class_name],
             PLUGIN_PATH_CACHE=PLUGIN_PATH_CACHE[self.class_name],
+            PLUGIN_INSTANCE_CACHE=PLUGIN_INSTANCE_CACHE[self.class_name],
```
this should be unnecessary if the above stuff is handled

s-hertel replied:
It is removed.

nitzmahone commented on lib/ansible/plugins/loader.py:
```diff
@@ -1056,6 +1086,15 @@ def all(self, *args, **kwargs):
             else:
                 fqcn = f"ansible.builtin.{basename}"
             self._update_object(obj, basename, path, resolved=fqcn)
+            if self._plugin_instance_cache is not None:
+                context = PluginLoadContext()
+                context.resolved = True
+                context.original_name = basename
+                context.plugin_resolved_name = basename
+                context.plugin_resolved_path = path
+                context.plugin_resolved_collection = 'ansible.builtin' if fqcn.startswith('ansible.builtin.') else ''
+                context._resolved_fqcn = fqcn
+                self._plugin_instance_cache[path] = (obj, context)
```
Distributing a copy/paste of this logic here doesn't feel right- `all()` is already duplicating a bunch of PluginLoader enumeration/instantiation logic as it is. I think you're on the right track with factoring out some of the `get_with_context` stuff into separate methods in this PR, so maybe `all` can piggyback on some of that to avoid duplicating the creation of the load context.

s-hertel replied:
I updated `all()` to call `get_with_context` to store the plugin in the cache and removed the external PluginLoader changes. The reason I made a separate cache using the path  as the key (as opposed to the plugin name) was to avoid the _resolve_step, which is slow. To work around that, `get_with_context` now can short-circuit before `_resolve_step` is called (for vars plugins only). I left a code comment about why that's there, in case we expand the cache to handle other plugin types, since that could be an issue for plugin types that use the collections list and need to be resolved every time.

nitzmahone commented on lib/ansible/plugins/vars/host_group_vars.py:
```diff
@@ -55,60 +55,92 @@
+            try:
+                if not entity.name or entity.base_type not in ('Host', 'Group'):
+                    raise AttributeError
+            except AttributeError:
+                raise AnsibleParserError("Supplied entity must be Host or Group, got %s instead" % (type(entity)))
+
+            entity_id = entity.base_type + entity.name
+            if entity_id in HOSTS or entity_id in GROUPS:
+                subdir = 'host_vars' if entity_id in HOSTS else 'group_vars'
+            elif entity.base_type == 'Host':
                 subdir = 'host_vars'
-            elif isinstance(entity, Group):
+                HOSTS.add(entity_id)
+            elif entity.base_type == 'Group':
                 subdir = 'group_vars'
-            else:
-                raise AnsibleParserError("Supplied entity must be Host or Group, got %s instead" % (type(entity)))
+                GROUPS.add(entity_id)
```
At a glance, I think we can lose all this stuff around the `HOSTS` and `GROUPS` caches entirely- if you reimplement the `Host` and `Group` `base_type` class vars as Enums as I described above, you can use the original logic as it is in `devel`, and just replace, eg, `isinstance(entity, Group)` with `entity.base_type is GroupEnumValue` (the `GROUP` value of the type enum I mentioned earlier).

s-hertel replied:
removed

nitzmahone commented on lib/ansible/plugins/vars/host_group_vars.py:
```diff
@@ -55,60 +55,92 @@
         data = {}
         for entity in entities:
-            if isinstance(entity, Host):
+            try:
+                if not entity.name or entity.base_type not in ('Host', 'Group'):
+                    raise AttributeError
+            except AttributeError:
+                raise AnsibleParserError("Supplied entity must be Host or Group, got %s instead" % (type(entity)))
```
The first half of the conditional could falsely trip on an empty string or `None` `entity.name`. Since we're in a really hot loop, and this will end up being a fatal error anyway, it's usually better to code for the "happy path" of the loop without doing defensive checks upfront, and just wrap the places you want to provide better error handling with tightly-scoped exception handlers instead (as they're much cheaper at runtime).

s-hertel replied:
I fixed it to make the exception handling more specific. An empty name or None should be an error. When we checked the entity was a Group/Host using `isinstance`, empty/null/non-string names were enforced at the inventory plugin level (assuming they are following the developer docs which will use [add_host,](https://github.com/ansible/ansible/blob/devel/lib/ansible/inventory/data.py#L191) and [add_group](https://github.com/ansible/ansible/blob/devel/lib/ansible/inventory/data.py#L160)).

nitzmahone commented on lib/ansible/plugins/vars/host_group_vars.py:
```diff
@@ -55,60 +55,92 @@
 
             # avoid 'chroot' type inventory hostnames /path/to/chroot
-            if not entity.name.startswith(os.path.sep):
+            if entity.name[0] != os.path.sep:
```
Yikes, I'd never really considered why this conditional is here- if we actually allow "pathy" looking things as host/group names by the time we've reached this point,  I suspect we might have a whole bunch of other "interesting" problems/bugs/potential-security-issues to solve...

bcoca replied:
the reason is chroot connection plugin

nitzmahone commented on lib/ansible/plugins/vars/host_group_vars.py:
```diff
@@ -55,60 +55,92 @@
-                    if cache and key in FOUND:
-                        found_files = FOUND[key]
+                    opath = os.path.join(realpath_basedir, subdir)
+                    b_opath = to_bytes(opath)
```
this can move below the `if cache` block so we only do it when we need to

s-hertel replied:
Good catch. This conversion was needless too, because it's used with methods that also accept unicode.

nitzmahone commented on lib/ansible/plugins/vars/host_group_vars.py:
```diff
@@ -55,60 +55,92 @@
+                    if cache:
+                        if opath in NAK:
+                            continue
+                        key = '%s.%s' % (entity.name, opath)
```
it's not immediately clear to me what's going on here with the mixing of paths and composite entity/path keys in `NAK`

s-hertel replied:
NAK just contains the path, I copied it from the prototype PR. I overlooked that the key was also checked for in NAK - I removed that. The composite key is used in the FOUND cache, which was pre-existing.

nitzmahone commented on lib/ansible/plugins/loader.py:
```diff
@@ -867,8 +874,15 @@ def get_with_context(self, name, *args, **kwargs):
 
+        return self.get_from_context(plugin_load_context, *args, **kwargs)
+
+    def get_from_context(self, plugin_load_context, *args, **kwargs):
```
I see the need to split some of the plugin construction from plugin location for specific cases where we're serving it from a cache, but pretty sure we don't want to turn all of this into public API- it's *really* special-purpose. I'm wondering if it makes more sense to just special case short-circuit instance-cached plugins internally early on in `get_with_context`, and special-case a tiny bit of `all` to internally remember it so we're not having to leak the specifics of yet another PluginLoader cache behavior out to the call sites...

s-hertel replied:
Fixed.

nitzmahone reviewed (changes requested):
A lot of the vars plugin caching storage and logic should probably get pushed down into our actual plugin- we don't want that to be part of the public API for other plugin instances to mess around with. Some of the other caching logic (`HOSTS`/`GROUPS`) can be removed entirely with just a small change to the type "hint" class attr we're now providing to use a singleton type (eg enum) that we can just quickly `is` to get what we need.

On the PluginLoader stuff, I think we want to hide as much of the details of this caching behavior as possible *inside* PluginLoader, so the call sites all retain their existing call patterns instead of introducing new public (or even private) APIs into PluginLoader to deal with the "it might come from a cache" cases.

nitzmahone reviewed (changes requested):
A lot of the vars plugin caching storage and logic should probably get pushed down into our actual plugin- we don't want that to be part of the public API for other plugin instances to mess around with. Some of the other caching logic (`HOSTS`/`GROUPS`) can be removed entirely with just a small change to the type "hint" class attr we're now providing to use a singleton type (eg enum) that we can just quickly `is` to get what we need.

On the PluginLoader stuff, I think we want to hide as much of the details of this caching behavior as possible *inside* PluginLoader, so the call sites all retain their existing call patterns instead of introducing new public (or even private) APIs into PluginLoader to deal with the "it might come from a cache" cases.

s-hertel committed:
Refactor to use the cache in existing plugin loader methods

Rename the attribute again

Refactor host_group_vars with requested changes

Make changelog a bugfixes fragment

Add a deprecation fragment for v2 vars plugins.

Add type hints

s-hertel committed:
unbreak group_vars

s-hertel commented on lib/ansible/plugins/vars/host_group_vars.py:
```diff
@@ -55,60 +55,99 @@
+                    if os.path.isdir(opath):
+                        self._display.debug("\tprocessing dir %s" % opath)
+                        FOUND[key] = found_files = loader.find_vars_files(opath, entity_name)
+                        # FOUND[key] = found_files
```
```suggestion

```

s-hertel committed:
Apply suggestions from code review

nitzmahone committed:
misc tweaks

* always cache instance by both requested and resolved FQ name
* add lru_cache to stage calculation to avoid repeated config consultation

nitzmahone committed:
handle KeyError from missing stage option
